src/libsyntax/parse/parser.rs

```diff
@@ -4070,6 +4088,9 @@ impl<'a> Parser<'a> {
                     TraitBoundModifier::None
                 };
                 bounds.push(TraitTyParamBound(poly_trait, modifier));
+                if has_parens {
+                    self.expect(&token::CloseDelim(token::Paren))?;
+                }
             } else {
                 break
```

Comment:
Oh wait, the `break` is actually wrong in presence of `(` or `?`.
`fn f<T: ?>(){}` in particular compiles on all Rust versions since 1.0 😆